File: frontend/src/services/stayService.ts
import api from './api';
import type {
  AvailableRoom,
  HotelSettingsRequest,
  HotelSettingsResponse,
  StayRequest,
  StayResponse,
} from '../types/stay.types';
import type { SpringPage } from '../types/page.types';

const BASE_PATH = '/api/v1/stays';

export const stayService = {
  getAllStays: async (page = 0, size = 20): Promise<SpringPage<StayResponse>> => {
    const response = await api.get<SpringPage<StayResponse>>(
      `${BASE_PATH}?page=${page}&size=${size}&sort=actualCheckInTime,desc`,
    );
    return response.data;
  },

  getStayById: async (id: string): Promise<StayResponse> => {
    const response = await api.get<StayResponse>(`${BASE_PATH}/${id}`);
    return response.data;
  },

  createStay: async (data: StayRequest): Promise<StayResponse> => {
    const response = await api.post<StayResponse>(BASE_PATH, data);
    return response.data;
  },

  checkOut: async (id: string): Promise<StayResponse> => {
    const response = await api.put<StayResponse>(`${BASE_PATH}/${id}/check-out`, {});
    return response.data;
  },

  retryInvoiceCreation: async (id: string): Promise<StayResponse> => {
    const response = await api.post<StayResponse>(`${BASE_PATH}/${id}/invoice/retry`, {});
    return response.data;
  },

  retryCheckoutEmail: async (id: string): Promise<StayResponse> => {
    const response = await api.post<StayResponse>(`${BASE_PATH}/${id}/checkout-email/retry`, {});
    return response.data;
  },

  getHotelSettings: async (): Promise<HotelSettingsResponse> => {
    const response = await api.get<HotelSettingsResponse>(`${BASE_PATH}/settings`);
    return response.data;
  },

  updateHotelSettings: async (data: HotelSettingsRequest): Promise<HotelSettingsResponse> => {
    const response = await api.put<HotelSettingsResponse>(`${BASE_PATH}/settings`, data);
    return response.data;
  },

  getLastCompletedStayForGuest: async (guestId: string): Promise<StayResponse | null> => {
    const response = await api.get<StayResponse>(`${BASE_PATH}/guest/${guestId}/latest`, {
      validateStatus: (s) => s === 200 || s === 204,
    });
    return response.status === 204 ? null : response.data;
  },

  getAvailableRooms: async (): Promise<AvailableRoom[]> => {
    const response = await api.get<{ content: AvailableRoom[] }>('/api/v1/rooms', {
      params: { size: 200 },
    });
    return (response.data.content ?? []).filter((r) => r.status === 'CLEAN');
  },
};

File: frontdesk-service/src/main/java/com/hotelpms/frontdesk/stays/service/StayService.java
package com.hotelpms.frontdesk.stays.service;

import com.hotelpms.frontdesk.stays.dto.GuestLastStayResponse;
import com.hotelpms.frontdesk.stays.dto.StayRequest;
import com.hotelpms.frontdesk.stays.dto.StayResponse;
import com.hotelpms.frontdesk.stays.dto.StaySummaryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.lang.NonNull;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Service interface for Stay operations.
 */
public interface StayService {

    /**
     * Creates a new stay (Check-in). Orchestrates calls to the rooms/reservations
     * domains (in-process) and guest-service/billing-service (Feign) to validate.
     *
     * @param request the stay request details
     * @return the created stay response
     */
    StayResponse checkIn(StayRequest request);

    /**
     * Performs a guest check-out. Updates stay status, marks the room as DIRTY,
     * and verifies the billing folio is PAID.
     *
     * @param stayId  the ID of the stay to check out
     * @param hotelId the authenticated hotel UUID; the stay must belong to it (T-STAY-04)
     * @return the updated stay response
     */
    StayResponse checkOut(@NonNull UUID stayId, @NonNull UUID hotelId);

    /**
     * Gets a stay by its ID, scoped to the authenticated hotel.
     *
     * @param id      the stay ID
     * @param hotelId the authenticated hotel UUID; the stay must belong to it (T-STAY-04)
     * @return the stay response
     */
    StayResponse getStayById(@NonNull UUID id, @NonNull UUID hotelId);

    /**
     * Retrieves a paginated list of all stays belonging to the authenticated hotel.
     * Supports standard Spring Data pagination query parameters.
     *
     * @param pageable the pagination and sorting parameters
     * @param hotelId  the authenticated hotel UUID (T-STAY-04)
     * @return a page of stay responses
     */
    Page<StayResponse> getAllStays(Pageable pageable, @NonNull UUID hotelId);

    /**
     * Retrieves all stays for a given reservation, scoped to the authenticated hotel.
     *
     * @param reservationId the reservation ID
     * @param hotelId       the authenticated hotel UUID; the reservation must belong to it (T-STAY-04)
     * @param pageable      the pagination and sorting parameters
     * @return list of stay responses
     */
    Page<StayResponse> getStaysByReservationId(@NonNull UUID reservationId, @NonNull UUID hotelId, Pageable pageable);

    /**
     * Returns the most recent completed stay for a guest within the caller's hotel, for
     * check-in form pre-filling. Scoped to {@code hotelId} (T-STAY-06, IDOR/cross-tenant
     * stay history leak) — a guest's stay at a different hotel must never be returned.
     * Verifies the guest profile is still active in guest-service before returning data
     * (fail-safe: returns empty if the profile was anonymised or the service is unavailable).
     *
     * @param guestId the guest UUID
     * @param hotelId the authenticated hotel UUID; only a stay owned by this hotel is returned
     * @return an Optional containing the last CHECKED_OUT stay, or empty if none or guest inactive
     */
    Optional<StayResponse> getLastCompletedStayForGuest(@NonNull UUID guestId, @NonNull UUID hotelId);

    /**
     * Returns the most recent check-in date for a guest within a hotel.
     * Called by the guest-service GDPR legal-hold guard (T-GST-05) to verify
     * whether the TULPS five-year retention obligation has expired before
     * anonymising a guest profile.
     *
     * @param guestId the guest UUID; must not be {@code null}
     * @param hotelId the hotel UUID; must not be {@code null}
     * @return a response containing whether stays exist and the most recent date
     */
    GuestLastStayResponse getLastStayDateForGuest(@NonNull UUID guestId, @NonNull UUID hotelId);

    /**
     * Returns all stay summaries for a guest within a hotel, ordered by check-in
     * descending. Used by the GDPR Art. 20 data-export endpoint.
     *
     * @param guestId the guest UUID; must not be {@code null}
     * @param hotelId the hotel UUID; must not be {@code null}
     * @return list of stay summaries, most recent first
     */
    List<StaySummaryResponse> getStayHistoryForGuest(@NonNull UUID guestId, @NonNull UUID hotelId);

    /**
     * Retries billing-invoice creation for a stay whose check-in-time attempt failed
     * (billing-service was unavailable). Clears {@code invoiceCreationFailed} on success.
     *
     * @param stayId  the stay ID
     * @param hotelId the authenticated hotel UUID; the stay must belong to it (T-STAY-04)
     * @return the updated stay response
     */
    StayResponse retryInvoiceCreation(@NonNull UUID stayId, @NonNull UUID hotelId);

    /**
     * Retries the checkout summary email for a checked-out stay whose original attempt
     * failed (notification-service was unavailable). Clears {@code checkoutEmailFailed}
     * on success.
     *
     * @param stayId  the stay ID
     * @param hotelId the authenticated hotel UUID; the stay must belong to it (T-STAY-04)
     * @return the updated stay response
     */
    StayResponse retryCheckoutEmail(@NonNull UUID stayId, @NonNull UUID hotelId);
}
